### ISSUE TITLE
KP_0 inside mode brackets does not work but outside it does.

### ISSUE THREAD

**Issue by freed00m** [i3/i3#2373]
Output of `i3 --moreversion 2>&- || i3 --version`:
```
Binary i3 version:  4.12-82-gaf1ab57 (2016-05-11, branch "gaps-next") © 2009 Michael Stapelberg and contributors
Running i3 version: 4.12-82-gaf1ab57 (2016-05-11, branch "gaps-next") (pid 686)
Loaded i3 config: /home/frdm/.i3/config (Last modified: Sat 11 Jun 2016 19:27:44 CEST, 619 seconds ago)

The i3 binary you just called: /usr/bin/i3
The i3 binary you are running: i3
```

URL to a logfile as per http://i3wm.org/docs/debugging.html:

[http://logs.i3wm.org/logs/5634208160546816.bz2](url)

**What I did:**

Tried to set xrandr option inside mode block brackets{ } and then running my long command  that works outside the brackets

```
bindsym XF86Display mode "$display_options"
set $display_options Choose xrandr output 
mode "$display_options" {
 #   Here it will not work and XRANRD will fail to set primary hdmi-0 and the nitrogen wont run either
 bindsym KP_0 exec --no-startup-id xrandr --output HDMI-0 --primary --mode 1920x1080 --pos 0x0 --rotate normal --output VGA-0 --off --output LVDS --off , mode "default", exec --no-startup-id nitrogen --restore
}


# But outside block will succeed with no problem
bindsym $mod+KP_0 exec --no-startup-id xrandr --output HDMI-0 --primary --mode 1920x1080 --pos 0x0 --rotate normal --output VGA-0 --off --output LVDS --off , mode "default", exec --no-startup-id nitrogen --restore
```

**What I saw:**

xrandr didn't passed all the parameters and didn't finish executing the whole line

**What I expected instead:**

execution inside the brackets { } to work exactly the same as outside the brackets

**Comment by Airblader** [comment]
From the i3 logfile i3 seems to be doing everything correct:

````
11/06/16 19:48:47 - commands_parser.c:parse_command:267 - COMMAND: *exec --no-startup-id xrandr --output HDMI-0 --primary --mode 1920x1080 --pos 0x0 --rotate normal --output VGA-0 --off --output LVDS --off , mode "default", exec --no-startup-id nitrogen --restore*                                                                                                                                                                                                                                                    
11/06/16 19:48:47 - commands.c:cmd_criteria_init:173 - Initializing criteria, current_match = 0x666c80
11/06/16 19:48:47 - commands.c:cmd_exec:1219 - should execute xrandr --output HDMI-0 --primary --mode 1920x1080 --pos 0x0 --rotate normal --output VGA-0 --off --output LVDS --off , no_startup_id = 1
11/06/16 19:48:47 - executing: xrandr --output HDMI-0 --primary --mode 1920x1080 --pos 0x0 --rotate normal --output VGA-0 --off --output LVDS --off 
11/06/16 19:48:47 - commands.c:cmd_mode:1019 - mode=default
11/06/16 19:48:47 - ../i3/src/bindings.c:switch_mode:470 - Switching to mode default
```

**Comment by Airblader** [comment]
@freed00m 

* How did you determine that part of the argument line is missing, opposed to something else being wrong?
* What happens if you put that long command into a separate script and call the script from your config? Does it work then? This is to rule out buffer issues (just in case).

**Comment by freed00m** [comment]
@Airblader Well, I determined that the resolution and the monitor is being switched but primary and everything after is dropped and even the nitrogen restore wont execute. And when I put the exact same command outside the block, everything including the nitrogen restore works as expected

**Comment by freed00m** [comment]
@Airblader If I put the command in a shell script and execute it via sh my_screen.sh, then it fails also

**Comment by Airblader** [comment]
@freed00m Is this reproducable 100% of the time for you or does it only happen sometimes?

According to the log, the nitrogen command was executed:

````
11/06/16 19:48:47 - commands.c:cmd_exec:1219 - should execute nitrogen --restore, no_startup_id = 1                                                                                                                                                           
11/06/16 19:48:47 - executing: nitrogen --restore
```

Can you adapt the command (mabye keep it in the external script for this) to redirect both stdout and stderr of the xrandr call to a file? Maybe `xrandr` tells us something about why it's failing.

**Comment by freed00m** [comment]
@Airblader It's happening everytime. During the log recording I have tried it twice, once with the outside block and the other with inside mode block. I will try stdout/stderr  tomorrow or the day after.

**Comment by Airblader** [comment]
> During the log recording I have tried it twice, once with the outside block and the other with inside mode block.

Ah, that changes things then. The log file only contains execution once, so the keybinding was simply never triggered in the other case.

Can you exchange the entire binding of the "inner" case with something like `exec i3-nagbar`? If it doesn't trigger that binding, maybe `KP_0` is the wrong keysym in this case. Have you verified with `xev -event keyboard` that `KP_0` is the keysym when not pressing the modifier key?

**Comment by freed00m** [comment]
Ok, so ther is a problem with KP_0

```
# Set position of HDMI_1920x1080
bindsym XF86Display mode "$display_options"
set $display_options Choose position of HDMI display
mode "$display_options" {
    bindsym KP_0 exec --no-startup-id i3-nagbar # Doesn't work
    bindsym KP_Decimal exec --no-startup-id i3-nagbar # this does work
}
```

Output when key is pressed

```
[frdm@normandy ~]$ xev -event keyboard
Outer window is 0x5400001, inner window is 0x5400002

KeymapNotify event, serial 24, synthetic NO, window 0x0,
    keys:  4294967230 0   0   0   16  0   0   0   0   0   0   0   0   0   0   0   
           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   

KeymapNotify event, serial 24, synthetic NO, window 0x0,
    keys:  4294967230 0   0   0   16  0   0   0   0   0   0   0   0   0   0[frdm@normandy ~]$ xev -event keyboard
Outer window is 0x5400001, inner window is 0x5400002

KeymapNotify event, serial 24, synthetic NO, window 0x0,
    keys:  4294967230 0   0   0   16  0   0   0   0   0   0   0   0   0   0   0   
           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   

KeyRelease event, serial 25, synthetic NO, window 0x5400001,
    root 0x499, subw 0x0, time 155652001, (-100,535), root:(815,564),
    state 0x10, keycode 36 (keysym 0xff0d, Return), same_screen YES,
"   XLookupString gives 1 bytes: (0d) "
    XFilterEvent returns: False

KeyPress event, serial 28, synthetic NO, window 0x5400001,
    root 0x499, subw 0x0, time 155656540, (-100,535), root:(815,564),
    state 0x10, keycode 90 (keysym 0xffb0, KP_0), same_screen YES,
    XLookupString gives 1 bytes: (30) "0"
    XmbLookupString gives 1 bytes: (30) "0"
    XFilterEvent returns: False

KeyRelease event, serial 28, synthetic NO, window 0x5400001,
    root 0x499, subw 0x0, time 155656591, (-100,535), root:(815,564),
    state 0x10, keycode 90 (keysym 0xffb0, KP_0), same_screen YES,
    XLookupString gives 1 bytes: (30) "0"
    XFilterEvent returns: False

KeymapNotify event, serial 28, synthetic NO, window 0x0,
    keys:  4294967193 0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
           0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
^C

```

**Comment by Airblader** [comment]
Yeah, but I too am a little confused over why it's not working. Maybe @stapelberg knows?

**Comment by stapelberg** [comment]
Is this a duplicate of #2346 by any chance?

**Comment by Airblader** [comment]
It sounds related, but I don't quite yet see that it'd be the same issue. That ticket is about KP_x being resolved in all cases. This ticket is about it not being resolved.

**Comment by stapelberg** [comment]
This is a duplicate of #2418 and hence fixed with commit  0469716fd6cbe22916b1f7a0cc8193f6025e8405.

I verified this using the following test case, which fails at commit 0469716fd6cbe22916b1f7a0cc8193f6025e8405^ and succeds at commit 0469716fd6cbe22916b1f7a0cc8193f6025e8405:

```perl
#!perl
# vim:ts=4:sw=4:expandtab
#
# Please read the following documents before working on tests:
# • http://build.i3wm.org/docs/testsuite.html
#   (or docs/testsuite)
#
# • http://build.i3wm.org/docs/lib-i3test.html
#   (alternatively: perldoc ./testcases/lib/i3test.pm)
#
# • http://build.i3wm.org/docs/ipc.html
#   (or docs/ipc)
#
# • http://onyxneon.com/books/modern_perl/modern_perl_a4.pdf
#   (unless you are already familiar with Perl)
#
# Verifies that one can bind on numpad keys in different numlock states.
# Ticket: #2346
# Bug still in: 4.12-78-g85bb324
use i3test i3_autostart => 0;
use i3test::XTEST;
use ExtUtils::PkgConfig;

SKIP: {
    skip "libxcb-xkb too old (need >= 1.11)", 1 unless
        ExtUtils::PkgConfig->atleast_version('xcb-xkb', '1.11');

my $config = <<EOT;
# i3 config file (v4)
font -misc-fixed-medium-r-normal--13-120-75-75-C-70-iso10646-1

mode "position" {
    bindsym KP_0 nop KP_0
}

bindsym Mod4+KP_0 nop Mod4+KP_0
EOT

my $pid = launch_with_config($config);

start_binding_capture;

is(listen_for_binding(
    sub {
        xtest_key_press(77); # enable Num_Lock
        xtest_key_release(77); # enable Num_Lock
        xtest_key_press(133); # Super_L
        xtest_key_press(90); # KP_0
        xtest_key_release(90); # KP_0
        xtest_key_release(133); # Super_L
    },
    ),
   'Mod4+KP_0',
   'triggered the "Mod4+KP_0" keybinding');

cmd 'mode "position"';

is(listen_for_binding(
    sub {
        xtest_key_press(90); # KP_0
        xtest_key_release(90); # KP_0
    },
    ),
   'KP_0',
   'triggered the "KP_0" keybinding');

sync_with_i3;
is(scalar @i3test::XTEST::binding_events, 2, 'Received exactly 2 binding events');

exit_gracefully($pid);

}

done_testing;
```
